torch/_decomp/decompositions.py

```diff
@@ -3249,6 +3249,96 @@ def upsample_bilinear2d(
     return result
 
 
+@register_decomposition(aten.replication_pad2d.default)
+@pw_cast_for_opmath
+def replication_pad2d(input: Tensor, padding: List[int]) -> Tensor:
+    pad_left = padding[0]
+    pad_right = padding[1]
+    pad_top = padding[2]
+    pad_bottom = padding[3]
+
+    # If all of the padding values are non-negative, then the following tensors
+    # are all equal to the input. But if any padding values are negative, we
+    # have to remove the appropriate rows and columns from the input.
+    # `input_mid` has all negative padding removed from it. `input_mid_tb` has
+    # negative left and right padding removed from it. `input_mid_lr` has
+    # negative top and bottom padding removed from it.
+    input_mid = input
+    input_mid_tb = input
+    input_mid_lr = input
+
+    if pad_left < 0:
+        input_mid = input_mid[..., -pad_left:]
+        input_mid_tb = input_mid_tb[..., -pad_left:]
+        pad_left = 0
+
+    if pad_right < 0:
+        input_mid = input_mid[..., :pad_right]
+        input_mid_tb = input_mid_tb[..., :pad_right]
+        pad_right = 0
+
+    if pad_top < 0:
+        input_mid = input_mid[..., -pad_top:, :]
+        input_mid_lr = input_mid_lr[..., -pad_top:, :]
+        pad_top = 0
+
+    if pad_bottom < 0:
+        input_mid = input_mid[..., :pad_bottom, :]
+        input_mid_lr = input_mid_lr[..., :pad_bottom, :]
+        pad_bottom = 0
+
+    batch_dims_no_repeat = (1,) * (input.dim() - 2)
+
+    repeat_top_left = batch_dims_no_repeat + (pad_top, pad_left)
+    repeat_top_middle = batch_dims_no_repeat + (pad_top, 1)
+    repeat_top_right = batch_dims_no_repeat + (pad_top, pad_right)
+
+    top_rows = torch.cat(
+        [
+            # top left
+            input[..., [0], :][..., [0]].repeat(repeat_top_left),
```

Comment:
I would be much happier if this code wasn't so hard coded for 2 dimensions, but the generated code looks surprisingly okay so I think this is fine for now.

Reply:
Yeah I was thinking about expanding it to work for 1d and 3d as well. I can do that in a follow up